src/compat.rs

```diff
@@ -50,17 +52,13 @@ where
             }
             assert!(!self.context.1.is_null());
             let waker = &mut *(self.context.1 as *mut _);
-            f(waker, Pin::new(&mut self.inner))
+            let mut guard = self.inner.lock();
+            match Pin::new_unchecked(&mut guard).poll(waker) {
```

Comment:
Not completely clear to me if this is safe to do, and if not what other options do we have?